### PR TITLE
[AIRFLOW-5221] add host_aliases to KubernetesPodOperator

### PR DESCRIPTION
**Pull request by kuromt**
add host_aliases paramter to KubernetesPodOperator.
The parameter is set to pod.spec.hostAliases.
Please see detail in [the kubernetes document](https://kubernetes.io/docs/concepts/services-networking/add-entries-to-pod-etc-hosts-with-host-aliases/).

---
Issue link: [AIRFLOW-5221](https://issues.apache.org/jira/browse/AIRFLOW-5221)

Make sure to mark the boxes below before creating PR: [x]

- [x] Description above provides context of the change
- [x] Commit message/PR title starts with `[AIRFLOW-NNNN]`. AIRFLOW-NNNN = JIRA ID<sup>*</sup>
- [x] Unit tests coverage for changes (not needed for documentation changes)
- [x] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [x] Relevant documentation is updated including usage instructions.
- [x] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

<sup>*</sup> For document-only changes commit message can start with `[AIRFLOW-XXXX]`.

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

### CODE REVIEW

File: airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py [hunk 1]
```diff
@@ -110,6 +110,8 @@ class KubernetesPodOperator(BaseOperator):  # pylint: disable=too-many-instance-
     :type is_delete_operator_pod: bool
     :param hostnetwork: If True enable host networking on the pod.
     :type hostnetwork: bool
+    :param host_aliases: A list of hostAliases.
+    :type host_aliases: list hostAliases
```
Comment (ashb) [inline airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py]:
```suggestion
    :type host_aliases: list
```

File: airflow/kubernetes/pod_generator.py [hunk 2]
```diff
@@ -116,6 +116,8 @@ class PodGenerator:
     :type affinity: Optional[dict]
     :param hostnetwork: If True enable host networking on the pod
     :type hostnetwork: bool
+    :param host_aliases: A list of hostAliases.
+    :type host_aliases: list hostAliases
```
Comment (ashb) [inline airflow/kubernetes/pod_generator.py:142]:
```suggestion
    :type host_aliases: list
```

Plus somewhere (not the type) a link to what structure this should have.

**Review by ashb — CHANGES_REQUESTED** [review]
Code okay.

@dimberman Do we want to carry on down this route or are we close to having your re-work of the entire kube executor config?

**Comment by kuromt** [comment]
```
No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#build-times-out-because-no-output-was-received
The build has been terminated
```

@ashb I merged your request, but It seems that travis has a problem.
Could you retry the building job?

**Comment by dimberman** [comment]
@ashb as the config rewrite is down the road I'm good to merge this.

**Comment by kuromt** [comment]
> @ashb as the config rewrite is down the road I'm good to merge this.

Thanks!

`pod_generator.py` is updated after my PR, so If I needed to rebase and commit again to merge this PR, please let me know.

**Comment by dimberman** [comment]
```
tests/runtime/kubernetes/test_kubernetes_executor.py .F

____ TestKubernetesExecutor.test_integration_run_dag_with_scheduler_failure ____

self = <tests.runtime.kubernetes.test_kubernetes_executor.TestKubernetesExecutor testMethod=test_integration_run_dag_with_scheduler_failure>

    def test_integration_run_dag_with_scheduler_failure(self):

        host = KUBERNETES_HOST

        dag_id = 'example_kubernetes_executor_config'

    

        result_json = self.start_dag(dag_id=dag_id, host=host)

    

        self.assertGreater(len(result_json['items']), 0)

    

        execution_date = result_json['items'][0]['execution_date']

        print("Found the job with execution date {}".format(execution_date))

    

        self._delete_airflow_pod()

    

        time.sleep(10)  # give time for pod to restart

    

        # Wait some time for the operator to complete

        self.monitor_task(host=host,

                          execution_date=execution_date,

                          dag_id=dag_id,

                          task_id='start_task',

                          expected_final_state='success', timeout=200)

    

        self.monitor_task(host=host,

                          execution_date=execution_date,

                          dag_id=dag_id,

                          task_id='other_namespace_task',

>                         expected_final_state='success', timeout=200)

tests/runtime/kubernetes/test_kubernetes_executor.py:231: 

_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

tests/runtime/kubernetes/test_kubernetes_executor.py:107: in monitor_task

    self.assertEqual(state, expected_final_state)

E   AssertionError: '' != 'success'

E   + success
```

**Comment by kuromt** [comment]
Thanks for your pointing.
Aftrer I rebased current apache/master  to this PR in local branch, tests are passed.

```
 pytest tests/runtime/kubernetes/test_kubernetes_executor.py 
================================================================================================= test session starts =================================================================================================
platform linux -- Python 3.6.10, pytest-5.4.1, py-1.8.1, pluggy-0.13.1 -- /usr/local/bin/python
cachedir: .pytest_cache
rootdir: /opt/airflow, inifile: pytest.ini
plugins: celery-4.4.2, flaky-3.6.1, cov-2.8.1, instafail-0.4.1.post0, requests-mock-1.7.0
collected 2 items                                                                                                                                                                                                     

tests/runtime/kubernetes/test_kubernetes_executor.py::TestKubernetesExecutor::test_integration_run_dag PASSED                                                                                                   [ 50%]
tests/runtime/kubernetes/test_kubernetes_executor.py::TestKubernetesExecutor::test_integration_run_dag_with_scheduler_failure PASSED                                                                            [100%]
```

**Comment by kuromt** [comment]
After extending timeout to 200, the test passed.
```
pytest tests/runtime/kubernetes/test_kubernetes_executor.py 
==================================================================== test session starts ====================================================================
platform linux -- Python 3.6.10, pytest-5.4.1, py-1.8.1, pluggy-0.13.1 -- /usr/local/bin/python
cachedir: .pytest_cache
rootdir: /opt/airflow, inifile: pytest.ini
plugins: celery-4.4.2, requests-mock-1.7.0, instafail-0.4.1.post0, cov-2.8.1, flaky-3.6.1
collected 2 items                                                                                                                                           

tests/runtime/kubernetes/test_kubernetes_executor.py::TestKubernetesExecutor::test_integration_run_dag PASSED                                         [ 50%]
tests/runtime/kubernetes/test_kubernetes_executor.py::TestKubernetesExecutor::test_integration_run_dag_with_scheduler_failure PASSED                  [100%]

=============================================================== 2 passed in 613.74s (0:10:13) ===============================================================
```

**Comment by kuromt** [comment]
The latest test failed due to timeout, but this fail is due to timing problem.
Should I kick off test again?

**Comment by jeffolsi** [comment]
this is very much needed. what is the status?

**Comment by dimberman** [comment]
Hi @kuromt could you please rebase with master? When those tests pass I'm good to merge

**Comment by kuromt** [comment]
@dimberman 

> Hi @kuromt could you please rebase with master? When those tests pass I'm good to merge

I rebase with master and pused to my branch.
But an integration test was canceled with `##[error]The operation was canceled.` message.

I don't have permission to retry this test. Can you retry this test?
https://github.com/apache/airflow/pull/7365/checks?check_run_id=680520307

**Comment by Ziliboba** [comment]
Hello, could you advice -  will these changes be merged any time soon? This functionality would be helpful.

**Comment by mik-laj** [comment]
I'm not a Kubernetes expert, but can't this option be configured with `pod` or `pod_template_file` parameter?
```
    :param pod: The fully specified pod. Mutually exclusive with `path_or_string`
    :type pod: Optional[kubernetes.client.models.V1Pod]
    :param pod_template_file: Path to YAML file. Mutually exclusive with `pod`
    :type pod_template_file: Optional[str]
```

**Comment by kuromt** [comment]
@mik-laj `pod_template_file` and `pod` are solutions for host_aliases. 
But host_aliases is useful because `pod_template_file` and `pod` parameters requires rest of whole complicate parameters of Pod and decreases flexibility.

**Comment by kuromt** [comment]
Today I run CI test but some tests failed.

I fix my bug and run CI test again.

**Comment by mik-laj** [comment]
@kuromt  Our mistake was to add a large number of parameters to this operator. This makes it very difficult to maintain.  We are now trying to change the approach and recommend using the above-mentioned parameters. We know that changing habits can be problematic for many users and there were even command that would facilitate migrations, but it is not finished yet.
https://github.com/apache/airflow/pull/8009
Would you like to help develop this command?  In my opinion, this is a better solution and more forward-looking. If we do not take steps in this direction, we will be sad all the time because we will be missing new Kubernetes features in Airfllow. We as project maintainers are also bored as we add/review every new feature the user may need.
I hope that the presented description of the situation will allow us to better understand the problem that we, as a community, face.

**Comment by kuromt** [comment]
@mik-laj Thanks. 
I agree that maintaing many parameters in KubernetesPodOperator is not a good way.

Let me confirm that the command is helper tool to generate API Object from existing dag files for `pod` or `pod_template_file` parameter, right?
I am interested in the idea.
To help developing the command, should I create new PR?

**Comment by mik-laj** [comment]
@kuromt  The new PR is a good solution.

**Comment by mik-laj** [comment]
@kuromt  Should we close this PR or do you want to keep working on it?

**Comment by kuromt** [comment]
@mik-laj Thaks, I'd like to close this PR.
@jeffolsi @Ziliboba Is that OK?

**Comment by kuromt** [comment]
OK, I close this PR.

@mik-laj Thanks for your support.

**Pull request closed without merging**